presto-spi/src/main/java/com/facebook/presto/spi/block/BlockDecoder.java

```diff
@@ -180,17 +125,117 @@ else if (map == null) {
         }
     }
 
+    static int[] getIdentityMap(int size)
+    {
+        return identityMapReference.updateAndGet(map -> {
+            if (map.length >= size) {
+                return map;
+            }
+            int[] newMap = Arrays.copyOf(map, size);
+            for (int i = map.length; i < newMap.length; i++) {
+                newMap[i] = i;
+            }
+            return newMap;
+        });
+    }
+
+    static int[] ensureCapacity(int[] map, int size)
+    {
+        if (map == null) {
+            return new int[size];
+        }
+        if (map.length >= size) {
+            return map;
+        }
+        return Arrays.copyOf(map, size);
+    }
+
     public void release()
     {
-        release(intArrayAllocator);
+        // TODO: what should the semantics of this be (if any).  Should it be close() to allow it
+        // to be nested in try-with-resources?  Do we need to pass in a handle to something?
+    }
+
+    private int[] getIntArray(int positionCount)
```

Comment:
What is this logic about? What is primary and secondary? Why is this method swaps them?

Reply:
The idea is to use ping-pong buffers to replicate the array reuse found in IntArrayAllocator.  There needs to be two of them because one is "hot" and is going have data written into it, and one will be cold and may have data being read from it (with a random access pattern).  I was trying to remove IntArrayAllocator and embed the minimum equivalent functionality directly in BlockDecoder.

That said, we still need to measure the benefit of BlockDecoder on the actual join code.  This class may not be necessary, we need to validate that it helps achieve the performance wins that we are expecting.